lib/rack/attack.rb

```diff
@@ -100,32 +147,87 @@ def initialize(app)
+      return @app.call(env) if !self.class.enabled || env["rack.attack.called"] || self.class.failure_cooldown?
 
-      env["rack.attack.called"] = true
+      env['rack.attack.called'] = true
       env['PATH_INFO'] = PathNormalizer.normalize_path(env['PATH_INFO'])
       request = Rack::Attack::Request.new(env)
+      result = :allow
 
+      self.class.with_calling do
+        result = get_result(request)
+      rescue StandardError => error
+        return do_error_response(error, request, env)
+      end
+
+      do_response(result, request, env)
+    end
+
+    private
+
+    def get_result(request)
       if configuration.safelisted?(request)
-        @app.call(env)
+        :allow
       elsif configuration.blocklisted?(request)
-        # Deprecated: Keeping blocklisted_response for backwards compatibility
-        if configuration.blocklisted_response
-          configuration.blocklisted_response.call(env)
-        else
-          configuration.blocklisted_responder.call(request)
-        end
+        :block
       elsif configuration.throttled?(request)
-        # Deprecated: Keeping throttled_response for backwards compatibility
-        if configuration.throttled_response
-          configuration.throttled_response.call(env)
-        else
-          configuration.throttled_responder.call(request)
-        end
+        :throttle
       else
         configuration.tracked?(request)
-        @app.call(env)
+        :allow
+      end
+    end
+
+    def do_response(result, request, env)
+      case result
+      when :block then do_block_response(request, env)
+      when :throttle then do_throttle_response(request, env)
+      else @app.call(env)
+      end
+    end
+
+    def do_block_response(request, env)
+      # Deprecated: Keeping blocklisted_response for backwards compatibility
+      if configuration.blocklisted_response
+        configuration.blocklisted_response.call(env)
+      else
+        configuration.blocklisted_responder.call(request)
+      end
+    end
+
+    def do_throttle_response(request, env)
+      # Deprecated: Keeping throttled_response for backwards compatibility
+      if configuration.throttled_response
+        configuration.throttled_response.call(env)
+      else
+        configuration.throttled_responder.call(request)
+      end
+    end
+
+    def do_error_response(error, request, env)
+      self.class.failed!
+      result = error_result(error, request, env)
+      result ? do_response(result, request, env) : raise(error)
+    end
+
+    def error_result(error, request, env)
+      handler = self.class.error_handler
+      if handler
+        error_handler_result(handler, error, request, env)
+      elsif self.class.allow_error?(error)
+        :allow
       end
     end
+
+    def error_handler_result(handler, error, request, env)
+      result = handler
+
+      if handler.is_a?(Proc)
+        args = [error, request, env].first(handler.arity)
```

Comment:
I've changed this so the handler has args `error, request` (since env is `request.env`). I've updated the Readme to show both args in the example.

I think we should keep the arity check here so you can do procs with 0 or 1 arg as a convenience.